### PR TITLE
GUI test and systems test suites

### LINKED ISSUE #162: Increase test coverage for files in .ui package
Currently the portion with the least coverage.
Will wait till @LeeEnHao is done before adding some tests.

### LINKED ISSUE #186: GUI tests for UI components
Migrate from previous module project

### LINKED ISSUE #189: Non GUI tests for UI components
References #162

### LINKED ISSUE #194: System tests according to specifications
Too much code. Not all features are finalised yet.

### PR DESCRIPTION
**Pull request by LeeEnHao**
## Description

All Ui tests and integrated systems tests.

Fixes #186
Fixes #189
Fixes #162 

## Testing

1. MainWindow has already been unit tested. To increase from 7 to 12 would be writing a full GUI systems test. Leaving it for a different PR.

1. Similarly UiManager testing would be a full system test. So leaving it to a different issue/PR.

1. ~~ClientCard is being tested with JVM modifying code. Client's final country field is being forcefully set to null to test the ClientCard does not accept null. Will remove tmr.~~

## Remarks

1. Left test for CountryNoteListPanel, will leave to another pr.

1. ~~Build file shorts GUI tests in GitHub environment.~~

1. TestFX [Documentation](https://testfx.github.io/TestFX/docs/javadoc/).

## Issues that needs to be resolved.

1. ~~Need to configure the build matrix of gradle.yml to properly setup the Linux environment to run headless test. To circumvent this, I am making `test` skip GUI tests in `CI` environment. Not exactly sure what the issue is but it seems xvfb needs to be downloaded. @qwoprocks any ideas 😕 ~~

## Acknowledgements

[AB4](https://github.com/se-edu/addressbook-level4), [#340](https://github.com/nus-cs2103-AY2021S1/forum/issues/340)

### CODE REVIEW

**Comment by LeeEnHao** [comment]
image

Ignore codecov.

File: src/test/java/guitests/guihandles/MainMenuHandle.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        clickOnMenuItemsSequentially("#help", "#helpMenuItem");
+    }
+
+    /**
+     * Opens the {@code HelpWindow} by pressing the shortcut key associated
+     * with the menu bar in {@code MainWindow}.
+     */
+    public void openHelpWindowUsingAccelerator() {
+        guiRobot.press(KeyCode.F1);
+    }
```
Comment (qwoprocks) [inline src/test/java/guitests/guihandles/MainMenuHandle.java]: These 2 functions don't seem to be used anymore

File: src/test/java/guitests/guihandles/StatusBarFooterHandle.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        lastRememberedSaveLocation = getSaveLocation();
+    }
+
+    /**
+     * Returns true if the current content of the 'save location' is different from the value remembered by the most
+     * recent {@code rememberSaveLocation()} call.
+     */
+    public boolean isSaveLocationChanged() {
+        return !lastRememberedSaveLocation.equals(getSaveLocation());
+    }
```
Comment (qwoprocks) [inline src/test/java/guitests/guihandles/StatusBarFooterHandle.java]: these functions dont seem to be used

File: src/test/java/seedu/address/ui/WidgetViewBoxTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        * Issue was raised by @rtshkmr where his JVM has correct timezone and time but wrong country.
+        * Locale.getDefault().getDisplayCountry() returns United States for him.
+        */
+        assumeFalse(guiRobot.isHeadlessMode(), "Test skipped in headless mode.");
+        WidgetViewBox widgetViewBox = new WidgetViewBox();
+
+        uiPartExtension.setUiPart(widgetViewBox);
+        guiRobot.pauseForHuman();
+        guiRobot.pauseForHuman();
+    }
```
Comment (qwoprocks) [inline src/test/java/seedu/address/ui/WidgetViewBoxTest.java]: Should this be paused for longer? I don't really notice it when I'm running the tests

File: src/test/java/seedu/address/ui/ClientListPanelTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            assertCardDisplaysClient(expectedClient, actualCard);
+            assertEquals(Integer.toString(i + 1) + ". ", actualCard.getId());
+        }
+    }
+
+    /*
+     * This is probably useless as TBM has almost no significant interaction with the ListView.
+     * However the this test is a good benchmark should we refactor the ListView. According to
+     * changes suggested in displaying notes, more performance test may be needed for this specific ListView.
+     */
```
Comment (qwoprocks) [inline src/test/java/seedu/address/ui/ClientListPanelTest.java]: Should add a TODO to remove this test at the end if we find that it's useless

**Review by qwoprocks — CHANGES_REQUESTED** [review]
Small changes

File: src/test/java/seedu/address/ui/HelpWindowTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        String ugUrl = (String) clipboard.getData(DataFlavor.stringFlavor);
+        assertEquals(USERGUIDE_URL, ugUrl);
+    }
+
+    @Test
+    public void copyUrlByDirectAccess_userGuideUrl_copiesCorrectly() throws Exception {
+        assumeFalse(guiRobot.isHeadlessMode());
+        // This test is to increase coverage. Somehow IDEA does not
+        // pickup the copyURL method from the previous test.
+        // Two tests are needed in fact.
```
Comment (qwoprocks) [inline src/test/java/seedu/address/ui/HelpWindowTest.java]: Should probably remove this comment lol

File: src/test/java/seedu/address/ui/CountryNoteCardTest.java [hunk 6]
```diff
@@ -29,6 +29,7 @@ private void assertCardDisplay(CountryNoteCard countryNoteCard, CountryNote expe
         guiRobot.pauseForHuman();
 
         CountryNoteCardHandle countryNoteCardHandle = new CountryNoteCardHandle(countryNoteCard.getRoot());
-        assertEquals(expectedId + ". " + expectedCard.getNoteContents(), countryNoteCardHandle.getCountryNoteContent());
+        assertEquals(expectedId + ". " + expectedCard.getNoteContents() + " [Singapore (SG)]",
```
Comment (qwoprocks) [inline src/test/java/seedu/address/ui/CountryNoteCardTest.java]:
```suggestion
        assertEquals(expectedId + ". " + expectedCard.getNoteContents() + " [" + expectedCard.getCountry() + "]",
```

**Review by qwoprocks — APPROVED** [review]
LGTM, except for the one small change

**Pull request merged**